catalog/core/catalog-core-commands/src/main/java/org/codice/ddf/commands/catalog/DuplicateCommands.java

```diff
@@ -140,41 +145,138 @@
      */
-    protected int queryAndIngest(CatalogFacade queryFacade, CatalogFacade ingestFacade,
-            int startIndex, Filter filter) {
+    protected void duplicateInBatches(CatalogFacade queryFacade, CatalogFacade ingestFacade,
+            Filter filter) throws Exception {
+        AtomicInteger queryIndex = new AtomicInteger(1);
 
-        // If maxMetacards is set, restrict the batchSize of the query to the remaining maxMetacards
-        if (maxMetacards > 0 && ((maxMetacards - successCount.get()) < batchSize)) {
-            batchSize = maxMetacards - successCount.get();
+        final long originalQuerySize;
+        if (maxMetacards > 0 && maxMetacards < batchSize) {
+            originalQuerySize = maxMetacards;
+        } else {
+            originalQuerySize = batchSize;
         }
 
-        List<Metacard> queryMetacards;
-        queryMetacards = query(queryFacade, startIndex, filter);
+        final SourceResponse originalResponse = query(queryFacade,
+                filter,
+                queryIndex.get(),
+                originalQuerySize);
+        if (originalResponse == null) {
+            return;
+        }
 
-        if (queryMetacards == null || queryMetacards.isEmpty()) {
-            return 0;
+        final long totalHits = originalResponse.getHits();
+        if (totalHits <= 0) {
+            return;
         }
 
-        List<Metacard> createdMetacards = ingestMetacards(ingestFacade, queryMetacards);
-        int failed = queryMetacards.size() - createdMetacards.size();
-        if (failed != 0) {
-            LOGGER.info("Not all records were ingested. [{}] failed", failed);
-            failedCount.addAndGet(failed);
-            failedMetacards.addAll(subtract(queryMetacards, createdMetacards));
+        // If the maxMetacards is set, restrict the totalWanted to the number of maxMetacards
+        final long totalWanted;
+        if (maxMetacards > 0 && maxMetacards <= totalHits) {
+            totalWanted = maxMetacards;
+        } else {
+            totalWanted = totalHits;
+        }
+
+        ingestMetacards(ingestFacade, getMetacardsFromSourceResponse(originalResponse));
+
+        if (multithreaded > 1) {
+            BlockingQueue<Runnable> blockingQueue = new ArrayBlockingQueue<>(multithreaded);
+            RejectedExecutionHandler rejectedExecutionHandler =
+                    new ThreadPoolExecutor.CallerRunsPolicy();
+            final ExecutorService executorService = new ThreadPoolExecutor(multithreaded,
+                    multithreaded,
+                    0L,
+                    TimeUnit.MILLISECONDS,
+                    blockingQueue,
+                    rejectedExecutionHandler);
+            console.printf("Running a maximum of %d threads during replication.%n", multithreaded);
+
+            printProgressAndFlush(start, totalWanted, ingestedCount.get());
+            int index;
+            while ((index = queryIndex.addAndGet(batchSize)) <= totalWanted) {
```

Comment:
what if most of the records failed?  your query index could be 28001, but you may have only successfully gotten 39 metacards... but this will quit after you queried for total # wanted, not when you have received the total # wanted or reached the end of what is available.